src/v/datalake/partition_spec_parser.cc

```diff
@@ -10,260 +10,516 @@
 
 #include "datalake/partition_spec_parser.h"
 
+#include "container/fragmented_vector.h"
+#include "iceberg/transform.h"
+#include "iceberg/unresolved_partition_spec.h"
+#include "utils/fixed_string.h"
+#include "utils/variant.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <fmt/format.h>
+
+#include <charconv>
+#include <optional>
+#include <string_view>
+#include <unordered_set>
+
 namespace datalake {
 
 namespace {
 
-template<typename T>
-struct parse_result {
-    T val;
-    std::string_view unparsed;
-};
+using source = std::string_view;
+using formatted_error = ss::sstring;
 
-struct parse_ctx {
-    std::string_view original;
-    ss::sstring last_error;
+class parser_base {
+protected:
+    explicit parser_base(source original)
+      : _original(original) {}
 
-    void report_expected(std::string_view unparsed, std::string_view expected) {
-        last_error = fmt::format(
-          "col {}: expected {} (got instead: `{}')",
-          unparsed.data() - original.data(),
-          expected,
-          unparsed);
-    }
-};
+    using position = source::const_iterator;
+    using expectation = std::string_view;
+    using expectations = std::unordered_set<expectation>;
+
+    template<typename T>
+    struct parse_success {
+        T val;
+        position next;
+    };
+
+    template<typename T>
+    class parse_result {
+        std::optional<parse_success<T>> res;
+
+        parse_result() = default;
+
+    public:
+        // success
+        parse_result(T&& val, position next)
+          : res(std::in_place, std::move(val), next) {}
+
+        // success
+        parse_result(
+          T&& val,
+          position next,
+          parser_base& self,
+          expectation exp,
+          position err_pos)
+          : res(std::in_place, std::move(val), next) {
```

Comment:
did you intend to forward from val?

Reply:
I did not intend to instantiate `parse_result` or `parse_success` templates with a reference-type `T` TBH. It may make sense in some weird situation, e.g. where the result of parsing is a reference to something present before parsing and outliving it. So I'll change it to forward.